### ISSUE TITLE
Eventflow stops periodical output

### ISSUE THREAD

**Issue by NicolasPrivatViseo** [Azure/diagnostics-eventflow#311]
Hi,
We have a Service Fabric application which uses eventflow to send logs into an event hub and appInsights (and a HttpOuput for debug purpose).
At startup event are correctly sent in the outputs but after a while (not clearly defined) it stops outputting the events.

We dug into the code and saw that you are using TPL Dataflow with BatchBlock. We also noticed the timer mecanism to trigger the batch in the DiagnosticPipeline initialization. 
We noticed that if we reach the batch size (100 events), then eventflow sends the output.
But the timer (`batchTimer` in `DiagnosticPipeline.cs` class) should allow to send the events even if the batch is not completed as it works at startup. 
The timer is configured with the default 500ms delay. The timer still executes the function `TriggerBatch` but the events are not sent to outputs.

As we can see in the capture below, we have no postponed message, and 10 events in the input queue, that are not sent in the output queue of the `BatchBlock`

![image](https://user-images.githubusercontent.com/44289644/54547961-1f990980-49a7-11e9-9cfc-c2164a091c9e.png)

Have you encountered similar issue, and do you have advice on where to look at?

We are using .Net Core 2.1, we have cloned the eventflow core (master) repo locally to debug within our project.

**Comment by karolz-ms** [comment]
Hi @NicolasPrivatViseo 

thanks for reporting this issue. This is first time we hear about it.

From what you are saying I assume that the EventFlow timer keeps calling `TriggerBatch` but, for some reason, the batch is not being actually triggered?

If so, it would be interesting to look at `BatchBlock` more closely. In particular, when you get to the state described above, could you check whether `_dataflowBlockOptions.CancellationToken.IsCancellationRequested` is true and what [`_target.BatchesNeedProcessing`](https://github.com/dotnet/corefx/blob/master/src/System.Threading.Tasks.Dataflow/src/Blocks/BatchBlock.cs#L570) property returns?

**Comment by NicolasPrivatViseo** [comment]
Hi @karolz-ms,
Thanks for the quick answer. Indeed the function `TriggerBatch` keeps being triggered.

It calls the [ProcessAsyncIfNecessary](https://github.com/dotnet/corefx/blob/master/src/System.Threading.Tasks.Dataflow/src/Blocks/BatchBlock.cs#L623) method. The `BatchesNeedProcessing` property is false.

About `_dataflowBlockOptions.CancellationToken.IsCancellationRequested`, it is set to false 

![image](https://user-images.githubusercontent.com/44289644/54594912-e78ad880-4a31-11e9-9561-a94fef2415f0.png)

We use debug symbols from Microsoft Symbol Server and have activated the Source Link Support option, but we have some trouble to debug properly with that (not all breakpoints can be set).

**Comment by karolz-ms** [comment]
That would explain why `TriggerBatch` has no effect, but we need to understand why.

You might be able to set a breakpoint on the BatchesNeedProcessing property getter and when you "step into", examine the local variables. That might give us a clue why the block thinks no processing is necessary.

@stephentoub any other ideas how to debug this (`BatchBlock.TriggerBatch` has no effect despite items sitting in input queue?)

**Comment by NicolasPrivatViseo** [comment]
We went into the `BatchesNeedProcessing` to see what was happening. We couldn't step into each line because of optimizations of the library I guess, but we could figure out how it behaves. It always returns false because of the number of `_messages` and `postponedMessages` stays at zero:

![image](https://user-images.githubusercontent.com/44289644/54696168-a2e36800-4b2b-11e9-9963-4b1a3e44a122.png)

The events pushed into the `_messages` Queue seem to be consumed right away because each time a new event comes, the `_messages` queue is empty. So where those events go ?

**Comment by karolz-ms** [comment]
There is only one place where messages are removed from the `_messages` queue, that is [`MakeBatchIfPossible`](https://github.com/dotnet/corefx/blob/master/src/System.Threading.Tasks.Dataflow/src/Blocks/BatchBlock.cs#L747) You can try putting a breakpoint there and see when it is called and what happens.

`_messages` queue being empty is strange because in the first screenshot you posted `InputQueue` has 10 items... and that is basically `_messages.Count`...

I suppose there is no easy way for you do give me a standalone repro of the problem?

**Comment by NicolasPrivatViseo** [comment]
The events seem to be consumed, from what I can see in `MakeBatchIfPossible` the queue get emptied:

![image](https://user-images.githubusercontent.com/44289644/54767015-a266d200-4bfc-11e9-8d4a-9f14f77fc9ea.png)

The issue seems still further.
And I still have messages in the input queue of the BatchBlock not being consumed as shown in the capture of the original message.

And unfortunately no I don't have a standalone set up in an app to reproduce the behaviour.

**Comment by karolz-ms** [comment]
The only explanation I can offer is that in the last two screenshots you are looking at "good" state.: `TriggerBatch` is called, that causes `MakeBatchIfPossible` to create the batch, `_messages` queue is emptied, next time `BatchesNeedProcessing` is called it returns false because all messages are processed.

But in the first screenshot you are showing a different state, the "bad" one, when the messages linger in the input queue. Note that InputQueue in that screenshot is the same as [`_target._messages`](https://github.com/dotnet/corefx/blob/master/src/System.Threading.Tasks.Dataflow/src/Blocks/BatchBlock.cs#L1189). 

My suggestion would be to ensure that you get into the bad state, and only then examine  the details of `BatchesNeedProcessing` (I would expect it returns false, which isn't what should happen, but the key is to understand why).

And I appreciate your help here, very much so. I know this kind of "remote debugging" can be time-consuming and frustrating. So thank you.

**Comment by NicolasPrivatViseo** [comment]
We have indeed two states that are inconsistent between them regarding the object, but all my screenshots are from the state where the logs don't reach the outputs :( That's what I don't understand.

Thanks a lot for your help too, it guides us for what to look at.

**Comment by karolz-ms** [comment]
Honestly I am not sure how to investigate this further without being able to reproduce on my side. I am sorry. Please feel to re-open if you get any new information that might point us to why this is happening.

**Comment by karolz-ms** [comment]
@NicolasPrivatViseo just FYI, I spent a couple of days trying to reproduce the problem from the example you provided to Microsoft support. It happened only once. During that time what I observed was that the timer was not functioning properly (`TriggerBatch` was not called). I was not able to determine with 100% certainty what the root cause was. My current theory is we are experiencing some sort of initialization problem that leads to temporary threadpool thread exhaustion that in turn prevents the timer from firing properly.

I have a change (see above) that makes the timer usage a bit more robust, but to be frank, it is a bit a shot in the dark.

I also have a few suggestions for changing the code in `DeviceConnectionManager` service that might help avoid this issue:
1. You have methods such as `Init` and `InitConnections` that are called from asynchronous methods like `RunAsync`, but are not async themselves. And inside these methods you start some tasks, then do `Tasks.WaitAny(...).Wait()`. My suggestion would be to make `Init` etc. async methods instead, and just await on tasks launched inside them normally. This avoids blocking threadpool threads and keeps the threadpool task queue going.
1. In your code I have seen a method declared as `async void`. There is no good reason to do that in service code. If an async method does not return a value, just declare it as returning a `Task` (`async Task` instead of `async void`).

Hope this helps!
Karol
